File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-2.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int min = nums[0], max = nums[0];
        int minSum = 0, maxSum = 0, total = 0;
        for (int num: nums) {
            total += num;
            minSum = Math.min(minSum + num, num);
            maxSum = Math.max(maxSum + num, num);
            min = Math.min(min, minSum);
            max = Math.max(max, maxSum);
        }
        return max > 0 ? Math.max(max, total - min) : max;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.rs
impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let mut res = nums[0];
        let (mut cur_min, mut cur_max) = (1, 1);
        for &num in &nums {
            let tmp = cur_max * num;
            cur_max = (num * cur_max).max(num * cur_min).max(num);
            cur_min = tmp.min(num * cur_min).min(num);
            res = res.max(cur_max);
        }
        res
    }
}